lib/vector-core/src/enrichment_table/enrichment_tables.rs

```diff
@@ -0,0 +1,286 @@
+use super::EnrichmentTable;
+use arc_swap::ArcSwap;
+#[cfg(feature = "vrl")]
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::{collections::HashMap, sync::Mutex};
+
+/// The Enrichment Tables manages the collection of `EnrichmentTable`s loaded into Vector.
+/// Enrichment Tables go through two stages.
+///
+/// ## 1. Writing
+///
+/// The  tables are loaded. There are two elements that need loading. The first is the actual data.
+/// This is loaded at config load time, the actual loading is performed by the implementation of
+/// the `EnrichmentTable` trait. Next, the tables are passed through Vectors `Transform` components,
+/// particularly the `Remap` transform. These Transforms are able to determine which fields we will
+/// want to lookup whilst Vector is running. They can notify the tables of these fields no that the
+/// data can be indexed.
+///
+/// During this phase, the data needs to be mutated and shared around a number of potential
+/// different components. Also, performance is not absolutely critical at this phase. Thus, the
+/// data is stored within a `Mutex` - stored in the `loading` field of this struct.
+///
+/// ## 2. Reading
+///
+/// Once all the data has been loaded we can move to the next stage. This is signified by calling
+/// the `finish_load` method. At this point all the data is swapped out of the `Mutex` in the loading
+/// field into the `ArcSwap` of the `tables` field. `ArcSwap` provides lock-free read-only access
+/// to the data. From this point on we have fast, efficient read-only access and can no longer add
+/// indexes or otherwise mutate the data .
+///
+/// Cloning this object is designed to be cheap. The underlying data will be shared by all clones.
+///
+#[derive(Clone, Default)]
+pub struct EnrichmentTables {
+    loading: Arc<Mutex<Option<HashMap<String, Box<dyn EnrichmentTable + Send + Sync>>>>>,
+    tables: Arc<ArcSwap<Option<HashMap<String, Box<dyn EnrichmentTable + Send + Sync>>>>>,
+}
+
+impl EnrichmentTables {
+    pub fn new(tables: HashMap<String, Box<dyn EnrichmentTable + Send + Sync>>) -> Self {
```

Comment:
I'm curious if there's a performance gain to be had by moving the table types into a giant enum (which wouldn't be very giant, since we'll likely have < 10 table types for a long time), forgoing any dynamic dispatching.

Reply:
Do mean replacing the String here with an enum? That string is user provided, coming from the config - it's the name of the config section rather than the type. A user could provide two different csv sections, for example. Not sure how we could change that to an enum.

Reply:
No, I meant the `EnrichmentTable` trait object. See [this blog post](https://www.possiblerust.com/guide/enum-or-trait-object).